### PR TITLE
feat: Tunnel remote APIC port to local machine using Kurtosis Portal

### PR DESCRIPTION
**Pull request by gbouv**
## Description:
Automatically tunnel remote APIC port to local machine to enable seamless interactions with enclaves running in remote kurtosis backend

## Is this change user facing?
TBD

### CODE REVIEW

File: api/golang/go.mod [hunk 1]
```diff
@@ -1,13 +1,17 @@
 module github.com/kurtosis-tech/kurtosis/api/golang
 
-go 1.15
+go 1.18
```
Comment (gbouv) [inline api/golang/go.mod:3]: Discussed here: https://kurtosistech.slack.com/archives/C032ZB39AHH/p1679486198287329

File: .circleci/config.yml [hunk 2]
```diff
@@ -16,6 +16,12 @@ param_cli_cluster_backend: &param_cli_cluster_backend
     type: enum
     enum: ["docker"] # ["docker", "minikube"]
 
+# Install go version 1.18 to run our integration tests on the minimal version we support
+steps_install_go_1_18: &steps_install_go_1_18
+  - run: |
+      wget https://go.dev/dl/go1.18.10.linux-amd64.tar.gz
+      sudo rm -rf /usr/local/go
+      sudo tar -C /usr/local -xzf go1.18.10.linux-amd64.tar.gz
```
Comment (gbouv) [inline .circleci/config.yml:24]: our ubuntu_vm comes with go 1.17 by default. I could have upgraded the VM, but the version above this one comes with Go 1.20, and I figured the best is probably to test our code on the minimum version we support. So, installing it manually

File: container-engine-lib/lib/backend_impls/remote_context_backend/backend_creator.go [hunk 3]
```diff
@@ -89,7 +89,7 @@ func writeTlsConfigToTempDir(ca []byte, cert []byte, key []byte) (string, func()
 	if err = os.WriteFile(caAbsFileName, ca, tlsFilesPerm); err != nil {
 		return "", nil, stacktrace.Propagate(err, "Error writing content of CA to temporary file at '%s'", caAbsFileName)
 	}
-	certAbsFileName := path.Join(tempDirectory, caFileName)
+	certAbsFileName := path.Join(tempDirectory, certFileName)
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/remote_context_backend/backend_creator.go:92]: nice catch

File: api/golang/engine/lib/kurtosis_context/kurtosis_context.go [hunk 4]
```diff
@@ -60,12 +68,52 @@ func NewKurtosisContextFromLocalEngine() (*KurtosisContext, error) {
...
+					err,
+					"An error occurred creating a connection to the Kurtosis Portal Client at '%v'",
+					kurtosisPortalSocketStr,
+				)
+			}
+			portalClient = portal_api.NewKurtosisPortalClientClient(portalConn)
+
+			if _, err = portalClient.Ping(context.Background(), portal_constructors.NewPortalPing()); err != nil {
+				return nil, stacktrace.Propagate(err, "Unable to connect to the locally running Kurtosis portal daemon."+
+					" To use remote context, a Kurtosis Portal daemon must be running on this machine")
```
Comment (h4ck3rk3y) [inline api/golang/engine/lib/kurtosis_context/kurtosis_context.go:104]: can we point people to how we they can run it?

File: api/golang/engine/lib/kurtosis_context/kurtosis_context.go [hunk 5]
```diff
@@ -60,12 +68,52 @@ func NewKurtosisContextFromLocalEngine() (*KurtosisContext, error) {
...
+			}
+			return nil, nil
+		},
+	}
+	_, err = contexts_store_api.Visit[struct{}](currentContext, instantiatePortalClientVisitor)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Error building client to Kurtosis Portal daemon")
```
Comment (h4ck3rk3y) [inline api/golang/engine/lib/kurtosis_context/kurtosis_context.go]: s/client to/client for/g

File: api/golang/engine/lib/kurtosis_context/kurtosis_context.go [hunk 6]
```diff
@@ -323,8 +371,24 @@ func runReceiveStreamLogsFromTheServerRoutine(
 }
 
 func newEnclaveContextFromEnclaveInfo(
```
Comment (h4ck3rk3y) [inline api/golang/engine/lib/kurtosis_context/kurtosis_context.go:373]: hopefully there aren't many CLI commands left that talk to the backend directly to do APIC stuff.

File: api/golang/engine/lib/kurtosis_context/kurtosis_context.go [hunk 7]
```diff
@@ -323,8 +371,24 @@ func runReceiveStreamLogsFromTheServerRoutine(
...
+	ctx context.Context,
+	portalClient portal_api.KurtosisPortalClientClient,
 	enclaveInfo *kurtosis_engine_rpc_api_bindings.EnclaveInfo,
 ) (*enclaves.EnclaveContext, error) {
+	// for remote contexts, we need to tunnel the APIC port to the local machine
+	if portalClient != nil {
+		apicGrpcPort := enclaveInfo.GetApiContainerHostMachineInfo().GetGrpcPortOnHostMachine()
+		forwardApicPortArgs := portal_constructors.NewForwardPortArgs(apicGrpcPort, apicGrpcPort)
```
Comment (h4ck3rk3y) [inline api/golang/engine/lib/kurtosis_context/kurtosis_context.go:381]: is there a port clash that can happen here on the local machine?

**Review by h4ck3rk3y — APPROVED** [review]

File: api/golang/engine/lib/kurtosis_context/kurtosis_context.go (hunk 7 above)
Comment (gbouv) [reply api/golang/engine/lib/kurtosis_context/kurtosis_context.go:381]: Yes, very much, in which case the portal will throw an error which looks fairly expressive. We could bind it to a different port, but right now we would need to make sure both the CLI and the engine are aware of the new port it's running on. Let's see if that's an important blocker before we invest too much time in that. But I agree that's a risk I'm tracking

File: api/golang/engine/lib/kurtosis_context/kurtosis_context.go (hunk 6 above)
Comment (gbouv) [reply api/golang/engine/lib/kurtosis_context/kurtosis_context.go:373]: Yup, I looked around quickly, couldn't find any. Once all those PRs are merged, I'll do an extensive testing of the different commands on a remote context, that will surely catch this.

File: api/golang/engine/lib/kurtosis_context/kurtosis_context.go (hunk 4 above)
Comment (gbouv) [reply api/golang/engine/lib/kurtosis_context/kurtosis_context.go:104]: That's still a bit uncertain right now to be entirely honest with you 👼 👼 
So, until I upgrade the Portal to be runnable in a more productised way, I'll hold off adding more details here. But I can update once the portal run command looks better 👍

**Pull request merged**